v2/bigquery-to-elasticsearch/src/main/java/com/google/cloud/teleport/v2/templates/BigQueryToElasticsearch.java

```diff
@@ -0,0 +1,259 @@
+ * APP_ROOT=/path/to/app-root
+ * COMMAND_SPEC=/path/to/command-spec
+ *
+ * # Build and upload image
+ * mvn clean package \
+ * -Dimage=${TARGET_GCR_IMAGE} \
+ * -Dbase-container-image=${BASE_CONTAINER_IMAGE} \
+ * -Dbase-container-image.version=${BASE_CONTAINER_IMAGE_VERSION} \
+ * -Dapp-root=${APP_ROOT} \
+ * -Dcommand-spec=${COMMAND_SPEC}
+ *
+ * # Create an image spec in GCS that contains the path to the image
+ * {
+ *    "docker_template_spec": {
+ *       "docker_image": $TARGET_GCR_IMAGE
+ *     }
+ *  }
+ *
+ * # Execute template:
+ * API_ROOT_URL="https://dataflow.googleapis.com"
+ * TEMPLATES_LAUNCH_API="${API_ROOT_URL}/v1b3/projects/${PROJECT}/templates:launch"
+ * JOB_NAME="bigquery-to-elasticsearch-`date +%Y%m%d-%H%M%S-%N`"
+ *
+ * time curl -X POST -H "Content-Type: application/json"     \
+ *     -H "Authorization: Bearer $(gcloud auth print-access-token)" \
+ *     "${TEMPLATES_LAUNCH_API}"`
+ *     `"?validateOnly=false"`
+ *     `"&dynamicTemplate.gcsPath=${BUCKET_NAME}/path/to/image-spec"`
+ *     `"&dynamicTemplate.stagingLocation=${BUCKET_NAME}/staging" \
+ *     -d '
+ *      {
+ *       "jobName":"'$JOB_NAME'",
+ *       "parameters": {
+ *            "inputTableSpec":"'$TABLE'",
+ *            "nodeAddresses":"'$NODE_ADDRESSES'",
+ *            "index":"'$INDEX'",
+ *            "documentType":"'$DOCUMENT_TYPE'"
+ *        }
+ *       }
+ *      '
+ * </pre>
+ */
+public class BigQueryToElasticsearch {
+
+  /** The tag for the main output of the json transformation. */
+  static final TupleTag<FailsafeElement<TableRow, String>> TRANSFORM_OUT =
+      new TupleTag<FailsafeElement<TableRow, String>>() {};
+
+  /** The tag for the dead-letter output of the json to table row transform. */
+  static final TupleTag<FailsafeElement<TableRow, String>> TRANSFORM_DEADLETTER_OUT =
+      new TupleTag<FailsafeElement<TableRow, String>>() {};
+
+  /** The tag for the main output of the json transformation. */
+  static final TupleTag<FailsafeElement<TableRow, String>> UDF_OUT =
+      new TupleTag<FailsafeElement<TableRow, String>>() {};
+
+  /** The tag for the dead-letter output of the json to table row transform. */
+  static final TupleTag<FailsafeElement<TableRow, String>> UDF_TRANSFORM_DEADLETTER_OUT =
+      new TupleTag<FailsafeElement<TableRow, String>>() {};
+
+  /** String/String Coder for FailsafeElement. */
+  static final FailsafeElementCoder<String, String> FAILSAFE_ELEMENT_CODER =
+      FailsafeElementCoder.of(
+          NullableCoder.of(StringUtf8Coder.of()), NullableCoder.of(StringUtf8Coder.of()));
+
+  /** TableRow/String Coder for FailsafeElement. */
+  static final FailsafeElementCoder<TableRow, String> FAILSAFE_TABLE_ROW_ELEMENT_CODER =
+      FailsafeElementCoder.of(TableRowJsonCoder.of(), NullableCoder.of(StringUtf8Coder.of()));
+
+  /** The default suffix for error tables if dead letter table is not specified. */
+  private static final String DEFAULT_DEADLETTER_TABLE_SUFFIX = "_error_records";
```

Comment:
I am not sure how the UDF here would be useful given that the user can actually provide an ad-hoc sql query? Any field manipulations can be done within the SQL. Let me know if you think otherwise

Reply:
@sabhyankar I agree, it makes more sense to keep any manipulations that need to be done in the query for speed and simplicity.